### DISCUSSION TITLE
Separating router interface and implementation

### DISCUSSION THREAD

**Discussion by vjpr** [trpc/trpc#1003] [Q&A]
Say I have a monorepo with two clients using the same server. There is a dependency from the clients to the entire server implementation.

https://github.com/KATT/zart/blob/60e7b37445f442e3eafec22f24bb069ca076e461/apps/nextjs/src/pages/api/trpc/%5Btrpc%5D.ts#L5-L6

Is there a way to avoid a hard dependency on the router impl? In its defense though, because clients and servers live in same codebase this becomes less of an issue than it was traditionally.

GraphQL achieves this by using a schema, although I will admit that `graphql-code-generator` becomes clunky to use, and GQL is designed around language interoperability, whereas trpc focuses on everything as TS for now.

Just interested in thoughts around this.

**Comment by KATT** [comment]
The whole principle of tRPC is that it is schemaless with the schema being transient. It sort of requires you to have the dependent services in the same project so it's not really designed for public-facing APIs.

That said, there are discussions about it over at https://github.com/trpc/trpc/issues/755 and https://github.com/trpc/trpc/discussions/271
